### DISCUSSION TITLE
is there a quick addition and modification of elements, as in Windows forms?

### DISCUSSION THREAD

**Discussion by SaBond0** [AvaloniaUI/Avalonia#14207] [Q&A]
in Windows forms, you can easily and simply add and change the form and all its elements, is this possible in Avalonia?

**Comment by StefanKoell** [comment]
Can you provide more details? Usually you would want to handle the UI through the ViewModel but you also can add/change elements by code. Or are you talking about designer features?

**Reply by SaBond0** [reply]
here there is a panel of elements, let's say there is a button element in it, in Windows forms it could simply be dragged from the elements panel to the form and corrected and changed there without using the code, the code itself changed

**Reply by StefanKoell** [reply]
Ok, so you are referring to designer features. XAML based layouts work a bit different. There's no out of the box drag'n'drop form editor, like in WinForms. @wieslawsoltes did some work in this area: https://x.com/wieslawsoltes/status/1744303336788988009?s=20

**Reply by timunie** [reply]
no, we have no visual designer like WinForms or WPF has. We have a previewer. The layout is typically written in XAML. Check [Avalonia.Samples](https://github.com/AvaloniaUI/Avalonia.Samples) and [docs](https://docs.avaloniaui.net) for onboarding.

**Reply by thevortexcloud** [reply]
You also probably don't want a designer anyway. The WPF one does not layout code properly because it insists on using margins everywhere. You just have to handwrite it. Luckily it's fairly simple and is fairly similar to writing HTML.

**Reply by StefanKoell** [reply]
Totally agree! Coming from WinForms myself, it takes some time to get used to all the different layout controls but it really grows on you and you don't want to go back to a drag'n'drop designer. It's much faster and more convenient in the end.

**Reply by thevortexcloud** [reply]
Yeah, Winforms needs a designer because it's a code only system, so the designer is effectively a visual source generator. XAML on the other hand with a bit of practice, it's possible to easily look at it and see how its structured.

On a side note, Winforms also has layout controls. Most people just never used them which made their apps look terrible when looked at on any screen not the same size as the person who made the forms.

**Reply by SaBond0** [reply]
what are the layout controls?

**Reply by StefanKoell** [reply]
Check the help here:
https://docs.avaloniaui.net/docs/basics/user-interface/building-layouts/

**Reply by BoTech-Development** [reply]
Hi,

I've been developing a designer for Avalonia for about a month and yesterday I released the first version, [V1.1.0](https://github.com/BoTech-Development/BoTech.AvaloniaDesigner/releases/tag/V1.1.0).
The Designer is called [**BoTech.DesignerForAvalonia**](https://github.com/BoTech-Development/BoTech.AvaloniaDesigner) and you can find the open source repository [here](https://github.com/BoTech-Development/BoTech.AvaloniaDesigner).
This designer may not be exactly what you expected because it does not work pixel oriented. Although it was advised not to use a designer like the Windows Forms Designer, but mine works with the layout controls from Avalonia and it is intended to be used to assist you by creating views.

I hope, that I could answer your question.
